### PR TITLE
Replace enum's in API with int's

### PR DESCRIPTION
**Pull request by richsalz**
sanitytest ensures that this PR is both API and ABI compatible.

### CODE REVIEW

**Comment by kaduk** [comment]
I think I have said this before, but the sysv ABI spec for amd64 requires that enumerated types be 4 bytes (with a caveat that some compilers have an extension when larger values are given).
Is the team's concern that some supported platform does not have an ABI spec that makes a similar guarantee about the width of enumerated types?  (Is there a list of supported platforms?  Does someone need to go through all of their ABI specifications?)

**Comment by richsalz** [comment]
That's nice to know, but see test/sanitytest.c  we're covered.

**Comment by levitte** [comment]
I would like a public explanation, for example in the commit message. This is a 180° turn from what we have pushed for so far, and if there was a discussion previous to this PR, I haven't seen it (yet)

**Comment by richsalz** [comment]
So far there is team internal discussion going on.  I will add more text to the commit message now.  Someone had brought up that enum's aren't fixed size (emilia maybe) on the team list awhile ago.

File: include/openssl/ui.h [hunk 1]
```diff
@@ -266,14 +266,12 @@ DEFINE_STACK_OF(UI_STRING)
...
  * needed by method authors.
  */
-enum UI_string_types {
-    UIT_NONE = 0,
-    UIT_PROMPT,                 /* Prompt for a string */
-    UIT_VERIFY,                 /* Prompt for a string and verify */
-    UIT_BOOLEAN,                /* Prompt for a yes/no response */
-    UIT_INFO,                   /* Send info to the user */
-    UIT_ERROR                   /* Send an error message to the user */
-};
```
Comment (kroeckx) [inline include/openssl/ui.h]: This existed already in 1.0.2 and probably older.

Comment (richsalz) [inline include/openssl/ui.h]: but the comment sayss it was only for people writing their own UI method.

Comment (kroeckx) [inline include/openssl/ui.h]: Which includes a function returning "enum UI_string_types". This clearly needs to get reverted.

Comment (richsalz) [inline include/openssl/ui.h]: I disagree about the "clearly."  If that's the only gating item to getting approval, then I will (sadly) do it.  But we're not they're yet.

Comment (richsalz) [inline include/openssl/ui.h]: And, the only people who would use that function where those people writing their own UI method.  So I think that user population is zero.

Comment (kroeckx) [inline include/openssl/ui.h]: If you think nobody is using it, I suggest you remove the whole thing instead, but not in 1.1.1

Comment (richsalz) [inline include/openssl/ui.h]: It's on my list for a future api-breaking release.  We made things opaque knowing that we'd miss things and we'd fix it as bugs.  This is a little- to not-used API that I want to ABI-compatibly make safer.

**Comment by petrovr** [comment]
Rich Salz wrote:
> richsalz commented on this pull request.
>
>
> And, the only people who would use that function where those people writing their own UI method.  So I think that user population is zero.
Please do not remove UI_get_string_type! Population is not zero!

File: include/openssl/ui.h (hunk 1 above)
Comment (levitte) [inline include/openssl/ui.h]: Uhmmm...  We have a UI_METHOD in `apps/apps.c`.  I know that engine-pkcs11 have a variant of the same thing in one of their test programs.  I know Windows folks that have written variants that prompt for a password using a dialog box (that's actually the basis for its design).

**Comment by richsalz** [comment]
It is not being removed.  I want to change it to an int; would that be a problem?

File: include/openssl/ui.h (hunk 1 above)
Comment (richsalz) [inline include/openssl/ui.h]: and would any of that code break if it became an int?

**Comment by levitte** [comment]
I'm guessing that what starts scaring people is @kroeckx' suggestion to remove the whole thing.

File: include/openssl/ui.h (hunk 1 above)
Comment (kroeckx) [inline include/openssl/ui.h]: Changing it from the enum to an int *will* break applications using it, and it's been made clear that there are applications using it.

Comment (richsalz) [inline include/openssl/ui.h]: As I said, if necessary I'll remove it.  But if most of the uses are like the one in apps/apps.c (a switch on ```UI_get_string_type```) then *no code changes are necessary* -- apps.c isn't in this PR.

**Comment by petrovr** [comment]
Rich Salz wrote:
> richsalz commented on this pull request.
>
> As I said, if necessary I'll remove it.  But if most of the uses are like the one in apps/apps.c (a switch on ```UI_get_string_type```) then *no code changes are necessary* -- apps.c isn't in this PR.

Openssl apps.c implements case with default password, i.e. one per 
function. Usual implementation is with two cases - "default password" 
and custom implementation.
Actually second case is hidden in apps.c with call of UI_OpenSSL methods.

Roumen

**Comment by levitte** [comment]
I've had the time to listen and think, and my conclusion is that as this stands in its entirety, it's an incompatible API change, at least in parts.  It feels like this is pushing us toward a slippery slope of "what can we get away with" that's turing unsafe.  This needs to go back the drawing board.

-1

**Comment by levitte** [comment]
(my last comment removed)

**Comment by bernd-edlinger** [comment]
I think this is no good idea for 1.1.0 because changing an enum to int is itself an ABI change.

For instance on arm-none-eabi gcc uses -fshort-enums by default, which means
that an enum is only as wide as the largest value.

However on arm-none-gnueabihf the enums are always min, 4 bytes wide.

**Comment by richsalz** [comment]
Team is not interested.

**Pull request closed without merging**